linkerd/app/outbound/src/sidecar.rs

```diff
@@ -31,22 +32,49 @@ struct HttpSidecar {
     routes: watch::Receiver<http::Routes>,
 }
 
+#[derive(Clone, Debug)]
+struct Target<T>(T);
+
 // === impl Outbound ===
 
 impl Outbound<()> {
-    pub fn mk_sidecar<T, I, P, R>(&self, profiles: P, resolve: R) -> svc::ArcNewTcp<T, I>
+    pub fn mk_sidecar<T, I, R>(
+        &self,
+        profiles: impl profiles::GetProfile<Error = Error>,
+        policies: impl policy::GetPolicy,
+        resolve: R,
+    ) -> svc::ArcNewTcp<T, I>
     where
         // Target describing an outbound connection.
         T: svc::Param<OrigDstAddr>,
         T: Clone + Send + Sync + 'static,
         // Server-side socket.
         I: io::AsyncRead + io::AsyncWrite + io::Peek + io::PeerAddr,
         I: Debug + Unpin + Send + Sync + 'static,
-        // Route discovery.
-        P: profiles::GetProfile<Error = Error>,
         // Endpoint resolver.
         R: Resolve<ConcreteAddr, Endpoint = Metadata, Error = Error>,
     {
+        let discover = svc::mk(move |addr: discover::TargetAddr| {
+            let profile = profiles
+                .clone()
+                .get_profile(profiles::LookupAddr(addr.clone().0));
+            let policy = policies.get_policy(addr);
+            Box::pin(async move {
+                let (profile, policy) = tokio::join!(profile, policy);
+                // TODO(eliza): we already recover on errors elsewhere in the
+                // stack...this is kinda unfortunate...
+                let profile = profile.unwrap_or_else(|error| {
+                    tracing::warn!(%error, "Failed to resolve profile");
+                    None
+                });
+                let policy = policy.unwrap_or_else(|error| {
+                    tracing::warn!(%error, "Failed to resolve client policy");
+                    None
+                });
```

Comment:
I don't think policies should be optional. Errors should be propagated.

Reply:
If the error is `NotFound`, the policy controller is telling us that the thing we looked up is not something it knows about (i.e. not a Service IP, currently). So that case, at least, _should_ produce a `None`. I think that logic will have to be pushed down into the discovery service though, so that any error that makes it this far can bubble up.

Reply:
I don't think we should rely on that `NotFound` behavior. The controller should always provide a route if the proxy should route. Otherwise the proxy should fail.

Reply:
Okay, that means we'll want to change the current controller behavior to return a `Forward` backend for any IP it doesn't know about, then (cc @adleong)